src/language.cpp

```diff
@@ -369,93 +368,58 @@ void LanguageBundle::InitMetaInfo(int index)
 	for (uint32 i = 0; i < nrstrings; ++i) {
 		int size = this->values.at(i).size();
 		if (_language.GetStringName(i) == nullptr && size != 0) {
-			fprintf(stderr, "Language %s has a string at undefined index %u.\n",
+			error("Language %s has a string at undefined index %u.\n",
 					this->metadata->name, i);
-			exit(1);
 		}
 		if (size > 1 && size != this->metadata->nplurals) {
-			fprintf(stderr, "Language %s has %d plurals, but string '%s' has %d.\n",
+			error("Language %s has %d plurals, but string '%s' has %d.\n",
 					this->metadata->name, this->metadata->nplurals, _language.GetStringName(i), size);
-			exit(1);
 		}
 	}
 }
 
 /**
- * Converts a double value into a utf-8 string with the appropriate separators.
- * @param dest [out] A provided buffer to write the output into.
- * @param size The size of the provided buffer.
- * @param amt The double value to be converted.
+ * Converts a money value into a utf-8 string with the appropriate separators.
+ * @param money The value in cents to be converted.
+ * @return Formatted money string.
  */
-static void MoneyStrFmt(char *dest, size_t size, double amt)
+static std::string MoneyStrFmt(int64 money)
 {
-	const std::string curr_sym = _language.GetSgText(GUI_MONEY_CURRENCY_SYMBOL);
-	size_t curr_sym_len = curr_sym.size();
+	std::string result;
+	if (money < 0) {
+		result += "\u2212";  // U+2212 Unicode Minus sign.
+		money *= -1;
+	}
 
 	const std::string tho_sep  = _language.GetSgText(GUI_MONEY_THOUSANDS_SEPARATOR);
-	size_t tho_sep_len  = tho_sep.size();
-
-	const std::string dec_sep  = _language.GetSgText(GUI_MONEY_DECIMAL_SEPARATOR);
-	size_t dec_sep_len  = dec_sep.size();
-
-	std::unique_ptr<char[]> buf(new char[size]);
 
-	/* Convert double into a numeric string (with '.' as the decimal separator). */
-	uint len = snprintf(buf.get(), size, "%.2f", amt);
+	std::string before_thousands = std::to_string(money / 100);
+	int nr_digits = before_thousands.size();
 
-	/* Compute the offset of where we should begin counting for thousand separators. */
-	uint comma_start = (len - 3) % 3;
+	result += _language.GetSgText(GUI_MONEY_CURRENCY_SYMBOL);
 
-	/* What is the max number of characters we might append in a single loop?
-	 * This is used to prevent a potential buffer overflow. */
-	/* max_append_size = 'a "special" symbol plus a single digit [0-9]'. */
-	int max_append_size = std::max(std::max(curr_sym_len, tho_sep_len), dec_sep_len) + 1;
-
-	uint j = 0;
-	uint curpos = 0;
-
-	/* Also has the added bonus of 'removing' the
-	 * automagically included '-' symbol by snprintf */
-	if (amt < 0) {
-		dest[j++] = '-';
-		curpos++;
+	for (int d = 0; d < nr_digits; ++d) {
+		result += before_thousands.at(d);
+		if (d > 0 && d + 1 < nr_digits && (nr_digits - d) % 3 == 1) result += tho_sep;
```

Comment:
If you don't want the `before_thousands` temporary, you could divide by 100 (and keeping it around), then recursively divide by 1000 until you reach the end. Then add currency symbol, optional minus sign, and unfold the recursion, appending the intermediate triplets to the string with a thousands separator between them.

Then finally add a decimal dot, and the remainder of the 100 division.

Not sure if you can do this easily, but printf formatter command `%03d` automatically prefixes zeroes upto the specifed length.

That said, this seem ok too.